### ISSUE TITLE
Optimise single-record calls to SeqIO.write and SeqRecord format

### ISSUE THREAD

**Issue by peterjc** [biopython/biopython#1097]
The design of ``SeqIO.write(...)`` intended it to be called once to write an entire file in one go, and indeed this is essential for complex file formats with a header/footer structure (e.g. XML). 

However, for many sequence file formats (e.g. FASTA, GenBank, EMBL), you can call ``SeqIO.write(single_record, output_handle, format)`` safely within a loop. However, it is much slower than it needs to be because it creates a writer object, calls the (dummy) header and footer writing code etc.

We can avoid this and at the same time also speed up the ``SeqRecord`` object's ``.format(...)`` method by special casing these file formats. For FASTA, GenBank, EMBL etc all we need is a dictionary of functions which take a single ``SeqRecord`` and return a string.

This shouldn't be that much work, just a lot of refactoring - moving the record writing code out of the existing FASTA/GenBank/Embl writer objects (which could then be deprecated) into the new standalone functions instead.

e.g. Translation script using a generator expression:

```python
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord

in_file = "pectate_lyases_nt.fasta"
out_file = "pectate_lyases_aa.fasta"
table = 11  # NCBI genetic code table for bacteria

def translate_cds_record(record, table=11):
    return SeqRecord(id=record.id,
                     description=record.description + " (protein)",
                     seq=record.seq.translate(table, cds=True))

SeqIO.write((translate_cds_record(r) for r in SeqIO.parse(in_file, "fasta")),
            out_file, "fasta")
```

Equivalent script using a for loop with multiple calls to ``SeqIO.write``:

```python
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord

in_file = "pectate_lyases_nt.fasta"
out_file = "pectate_lyases_aa.fasta"
table = 11  # NCBI genetic code table for bacteria

with open(out_file, "w") as out_handle:
    for nt_record in SeqIO.parse(in_file, "fasta"):
        aa_seq = nt_record.seq.translate(table, cds=True)
        print(nt_record.seq.translate(table, cds=True))
        aa_record = SeqRecord(id=nt_record.id,
                              description=nt_record.description + " (protein)",
                              seq=aa_seq)
        SeqIO.write(aa_record, out_handle, "fasta")
```

Other than having to explicitly use an output handle (using a filename would simply over-right the same file for each record, leaving you at the end with a FASTA file containing the final record only), I think this is much more accessible for a novice (as you don't need a generator expression or other advanced construct like a generator function).

This issue is mainly to remind me to get round to this when I have time, having thought about it one and off at intervals.

This would also need benchmarking.

**Comment by peterjc** [comment]
Related to this, it would make sense to expose simple functions for:

- building a FASTA formatted string (including sequence line wrapping) from a title and sequence (string or Seq object)
- building a FASTQ formatted string from a title, sequence (string or Seq object), and ASCI quality string

This would be analogous to the low-level parsing functions ``SimpleFastaParser`` and ``FastqGeneralIterator``, and would be helpful in a number of situations include teaching where you want to postpone creating Seq and SeqRecord objects.

**Comment by peterjc** [comment]
I've been exploring a proof-of-principle which introduces functions named ``as_fasta``, ``as_fastq``, ``as_fastq_solexa`` etc.

However named, should these ``SeqRecord`` to formatted string functions allow optional arguments? Perhaps only when implemented by the writer objects they would replace (like line wrapping for FASTA or QUAL ), or focus on keeping things simple?

**Comment by mdehoon** [comment]
Perhaps, instead of introducing one function for each format, we can have a single method (maybe called "format") that takes the requested format as the argument?
So record.format("fasta"), record.format("fastq"), record.format("fastq_solexa").Then we can have additional optional arguments as needed.
-Michiel.


    On Tuesday, March 21, 2017 2:34 AM, Peter Cock <notifications@github.com> wrote:
 

 I've been exploring a proof-of-principle which introduces functions named as_fasta, as_fastq, as_fastq_solexa etc.However named, should these SeqRecord to formatted string functions allow optional arguments? Perhaps only when implemented by the writer objects they would replace (like line wrapping for FASTA or QUAL ), or focus on keeping things simple?—

**Comment by peterjc** [comment]
@mdehoon we already have a ``SeqRecord`` method, e.g. ``record.format("fasta")``. Currently it calls ``SeqIO.write(...)`` with a ``StringIO`` handle internally, which is a lot of overhead. This issue would speed this up enormously.

I'm not entirely sure if the one-function-per-format ought even to be part of the public API? The idea is they will replace the current one-writer-class-per-format.

Certainly I would focus on the ``SeqRecord`` method and ``SeqIO.write(...)`` as the main interface in the Tutorial etc.

**Comment by peterjc** [comment]
This was included in Biopython 1.72 (21 June 2018), belatedly closing issue.
